distutils/extension.py

```diff
@@ -106,12 +108,23 @@ def __init__(
     ):
         if not isinstance(name, str):
             raise AssertionError("'name' must be a string")  # noqa: TRY004
-        if not (
-            isinstance(sources, list)
-            and all(isinstance(v, (str, os.PathLike)) for v in sources)
-        ):
+
+        # we handle the string case first; though strings are iterable, we disallow them
+        if isinstance(sources, str):
+            raise AssertionError(  # noqa: TRY004
+                "'sources' must be an iterable of strings or PathLike objects, not a string"
+            )
+
+        # mow we check if it's iterable and contains valid types
+        try:
+            sources = list(sources)  # convert to list for consistency
+            if not all(isinstance(v, (str, os.PathLike)) for v in sources):
+                raise AssertionError(
+                    "All elements in 'sources' must be strings or PathLike objects"
+                )
+        except TypeError:
             raise AssertionError(
```

Comment:
Question for the distutils maintainers: I can't help but feel that `AssertionError` is misused here, shouldn't these all be `TypeError` ? (it's noqa'd but not explained)

Reply:
I have the same suggestion! :)

Reply:
Peeking at the history, these used to be actual `assert` statements (https://github.com/pypa/distutils/commit/7d70ca7551824bdf4950932bd5d68e8a982cf1d0), so my guess is that the `AssertionError`s were kept for backwards compatibility with code that might try to catch these errors.

Reply:
Additionally, what I infer from the commit message is that that they were added because in-line assert statements are completely removed from the bytecode if used with `-O`

Reply:
I suspect any Exception here is fine, so a TypeError should be fine (famous last words).

Reply:
For uniformity, I changed the `AssertionError` for both the name and the sources checks in efeb97c02684965d63e78eb9319458b0e8074f66 to be a `TypeError`.